### ISSUE TITLE
Fix nat erasure in LF

### ISSUE THREAD

**Issue by sofiafaro-da** [digital-asset/daml#11679]
Idea is to change nat erasure in a future version of LF. Let's call it version X.

In version X:

- `nat` is an erasable kind, so all kinds are erasable and we don't have this weird split between erasable & non-erasable kinds. 
- To preserve type-level `nat` information at runtime where needed, we introduce a "nat singleton" type `NatSingleton n` that has the primitive constructor `ENat @n : NatSingleton n` for all nat literals `n`. 
- We introduce Numeric primitives that use the new nat singleton where needed, and otherwise get the numeric scale directly from their input (where possible).

Migration from old LF (< X) to new LF (>= X) requires some kind of translation inside the speedy compiler, such that a non-erasable `nat` gets turned into the combination of an erasable `nat` and a nat singleton. Here's the translation I have in mind:

- In type (definitions), any nat variables and literals remain the same.
  - Record and variants types keep the same type parameters, even if they are `nat` kinded.
    E.g. `record R (n: nat) = { ... }` remains the same, only the field types need to be translated.
- A type of the form `forall (n: nat), T` must be translated to `forall (n: nat), NatSingleton n -> T`.
- An expression of the form `Λ (n: nat). e` must be translated to `Λ (n: nat).  λ (n_repr : NatSingleton n). e`
- An expression of the form `e @n` where `n` is a nat type variable must be translated to `e @n n_repr`
- An expression of the form `e @n` where `n` is a nat literal must be translated to `e @n (ENat @n)`.

(This relies on all nat-kinded types being either nat literals or nat type variables. Which is true, I think.)

This translation is simple enough that it might be doable inline by speedy. What happens in speedy, to make nats non-erasable, is effectively this translation. So the idea is actually to keep the *current interpretation* for old LF programs, and simplify the interpretation (by erasing `nat`) in new LF programs.

(Then there's a lot of other questions, like how to deal with the change in damlc, in daml-stdlib, etc. I assume if we can figure out how to change things in LF, changing the rest while maintaining backward compatibility will be possible, even if it involves lots of glue.)

**Comment by cocreature** [comment]
> In types, any nat variables and literals remain the same.

>    Record and variants types keep the same type parameters, even if they are nat kinded.
>    E.g. record R (n: nat) = { ... } remains the same, only the field types need to be translated.

Do you mean in type definitions here? I’m a bit confused by the fact that you start by saying they stay the same in types but then exclude field types.

**Comment by cocreature** [comment]
Sounds like we could make this work, thanks a lot for writing this up. In terms of tradeoffs, I’m somewhat unsure:

1. As long as we have to keep support for LF versions around that do not have this change it seems like this adds complexity in the damlc compiler, LF typechecker & speedy compiler.
2. From a maintenance pov, we simplify the LF typecheckers slightly and probably a bit more in the daml compiler. Afaict, the speedy compiler doesn’t seem to get meaningfully simpler afaict we just end up treating the singleton like we treat nat type variables atm.
3. From a user pov, we solve some edge cases where the daml compiler currently falls apart. See #9181 for an example. Afaik we never had any user reports for those types of errors so I’m not all that concerned but is a wart.

So overall, it seems like in the short (and that’s at least a few months if not years) term we add a bit of complexity to eventually kill a bit of complexity in the compiler and solve some issues which don’t seem to matter much to users.

Do you agree with that characterization?

If so, do you think it’s still worth pursuing this? To me, it seems like we might be better off spending the effort here on other cleanup to reduce complexity but I’m happy to be convinced it’s worth the effort.

**Comment by sofiafaro-da** [comment]
I agree. This probably isn't worth pursuing at the moment, but it may be useful in the future.

**Comment by sofiafaro-da** [comment]
> > In types, any nat variables and literals remain the same.
> 
> > Record and variants types keep the same type parameters, even if they are nat kinded.
> > E.g. record R (n: nat) = { ... } remains the same, only the field types need to be translated.
> 
> Do you mean in type definitions here? I’m a bit confused by the fact that you start by saying they stay the same in types but then exclude field types.

Yes, in the type definition. You have to "translate" the field types (in case there's a `forall (n: nat)`) in there.

**Comment by remyhaemmerle-da** [comment]
#18453 closes this issue.
